torch/utils/data/dataloader.py

```diff
@@ -670,10 +670,10 @@ def _reset(self, loader, first_iter=False):
     def _next_index(self):
         return next(self._sampler_iter)  # may raise StopIteration
 
-    def _next_data(self):
+    def _next_data(self) -> T_co:
```

Comment:
You can't really infer the type here. First of all, it's going to be a batch of data when `batch_size` is specified. Second, data would be collated into Tensor if possible. Third, with custom `collate_fn`, `DataLoader` has no idea what type of data can be returned.

Reply:
Thanks for bring thing up. I was worried something like this might come up but it wasn't obvious to me where on a first attempt.

Regarding `collate_fn`, I imagine there should be an implicit contract that it not return arbitrary values; however, it seems that can't be enforced with type hints according to the comment above the definition of [`_collate_fn_t`](https://github.com/pytorch/pytorch/blob/master/torch/utils/data/dataloader.py#L57).

Regarding the problems with batching you've mentioned, I could perhaps change the return type to something like `Union[T_co, List[T_co], Tensor]`?

Reply:
You can see the output type from `collate_fn_t` is `Any`. Then, there is a chance that a custom `collate_fn` would return data in type of anything.

If you really want to type `_next_data` function, the only option IMHO is `Any`.

Reply:
Right, my understanding though from the comment is that the only reason the return type of `collate_fn` is `Any` is because of limitations of Python type hints. Ideally, we would something more constrained than that, no?

Reply:
I think the return type of `collate_fn` may actually need to be `Any` given that users can create a custom `collate_fn` function and the return type of that doesn't necessarily have to relate to the input type in anyway.

What would a feasible constrained return type be? Any suggestion?

Reply:
I suppose I would ideally want it to return `T_co` or perhaps `Union[T_co, List[T_co], Tensor]`.

Reply:
I think just based on `default_collate`, the types such as `Mapping[T_co]`, `NamedTuple[T_co]`, `Sequence[T_co]` are also possible return types. Again, users may decide to return custom objects unrelated to the input type, which makes this difficult.

Reply:
I think just based on `default_collate`, the types such as `Mapping[T_co]`, `NamedTuple[T_co]`, `Sequence[T_co]` are also possible return types. Again, users may decide to return custom objects unrelated to the input type, which makes this difficult.

Reply:
Nonetheless we appreciate your PR and thanks for keeping an eye out for potential improvements!